aten/src/ATen/native/xnnpack/Convolution.cpp

```diff
@@ -0,0 +1,309 @@
+#ifdef USE_XNNPACK
+
+#include <ATen/cpp_custom_type_hack.h>
+#include <ATen/native/ConvUtils.h>
+#include <ATen/native/utils/ParamUtils.h>
+#include <ATen/native/xnnpack/Common.h>
+#include <ATen/native/xnnpack/Factory.h>
+
+namespace at {
+namespace native {
+namespace xnnpack {
+namespace internal {
+namespace convolution2d {
+
+struct Context final {
+  Operator convolution_op;
+
+  std::vector<int64_t> weight_size;
+  std::vector<int64_t> padding;
+  std::vector<int64_t> stride;
+  std::vector<int64_t> dilation;
+
+  static constexpr float kMin = -std::numeric_limits<float>::infinity();
+  static constexpr float kMax = std::numeric_limits<float>::infinity();
+};
+
+namespace {
+
+// Supports NHWC and NCHW FP32 convolutions with any valid
+//  - kernel size
+//  - padding
+//  - stride
+//  - dilation
+//  - grouping
+
+bool available(
+    const Tensor& weight,
+    const c10::optional<Tensor>& bias,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const int64_t groups,
+    const float output_min,
+    const float output_max) {
+         // XNNPACK
+  return xnnpack::internal::available() &&
+         // Weight
+         (4 == weight.ndimension()) &&
+         (weight.size(Layout::Filter::Height) > 0) &&
+         (weight.size(Layout::Filter::Width) > 0) &&
+         (c10::DeviceType::CPU == weight.device().type()) &&
+         (kFloat == weight.scalar_type()) &&
+         // Bias
+         ((bias && bias->defined()) ? ((1 == bias->ndimension()) &&
+                                      (c10::DeviceType::CPU == bias->device().type()) &&
+                                      (kFloat == bias->scalar_type()) &&
+                                      (weight.size(Layout::Filter::Output)) == bias->size(0))
+                                    : true) &&
+         // Padding
+         (padding[Layout::Parameter::Height] >= 0) &&
+         (padding[Layout::Parameter::Width] >= 0) &&
+         // Stride
+         (stride[Layout::Parameter::Height] > 0) &&
+         (stride[Layout::Parameter::Width] > 0) &&
+         // Dilation
+         (dilation[Layout::Parameter::Height] > 0) &&
+         (dilation[Layout::Parameter::Width] > 0) &&
+         // Groups
+         (groups > 0) &&
+         // Input
+         (weight.size(Layout::Filter::Input) > 0) &&
+         // Output
+         (weight.size(Layout::Filter::Output) > 0) &&
+         // Output - Groups
+         ((weight.size(Layout::Filter::Output) % groups) == 0) &&
+         // Output Min / Max
+         (output_max > output_min);
```

Comment:
Just a personal style, I would write this as

```
(output_max > output_min) &&
true;
```

so more conditions can be added with a green-only diff.